### DISCUSSION TITLE
Decision logic for whether extensions can be enabled or disabled per workspace

### DISCUSSION THREAD

**Discussion by jpogran** [microsoft/vscode-discussions#930] [Extension Development QnA] [answered]
I'm one of the maintainers for the [HashiCorp Terraform extension](https://marketplace.visualstudio.com/items?itemName=HashiCorp.terraform) and we've recently received an [issue](https://github.com/hashicorp/vscode-terraform/issues/1629) showing that our extension can't be enabled or disabled at the workspace level. We did not intend on preventing this particular feature and would like to figure out what is controlling it. I attempted to find somewhere that explains what controls this functionality, but could not find anything for the extension author perspective, just articles that explain that the functionality is available.

As the issue states, we went from one version this being available to the most recent version where it is not available. Knowing what we added in the latest release, I could only think of the addition of using secret storage with an authentication provider as the "net new" thing. The rest of the new features are similar, at least I think, to functionality in previous versions (REST API calls, settings, etc).

Is there somewhere that outlines what an extension author can do to allow per workspace enablement?

**Comment by tamuratak** [comment]
It seems that if your extension contributes an authentication provider, it cannot change enablement in workspace. If you think the logic is not appropriate, you had better open an issue at the VS Code repository.

https://github.com/microsoft/vscode/blob/6f7e102f08acbb75fbdbe4c052b425149f10518a/src/vs/workbench/services/extensionManagement/browser/extensionEnablementService.ts#L182-L189

```ts
    private throwErrorIfCannotChangeWorkspaceEnablement(extension: IExtension): void {
        if (!this.hasWorkspace) {
            throw new Error(localize('noWorkspace', "No workspace."));
        }
        if (isAuthenticationProviderExtension(extension.manifest)) {
            throw new Error(localize('cannot disable auth extension in workspace', "Cannot change enablement of {0} extension in workspace because it contributes authentication providers", extension.manifest.displayName || extension.identifier.id));
        }
    }
```

https://github.com/microsoft/vscode/blob/6f7e102f08acbb75fbdbe4c052b425149f10518a/src/vs/platform/extensions/common/extensions.ts#L472-L474

```ts
export function isAuthenticationProviderExtension(manifest: IExtensionManifest): boolean {
    return manifest.contributes && manifest.contributes.authentication ? manifest.contributes.authentication.length > 0 : false;
}
```

**Reply by jpogran** [reply]
Thanks for the reply @tamuratak, but I still don't understand why that would affect enablement. The code in the links you provide are for browser based extensions, which doesn't apply when I'm using a locally installed version on my desktop. I also see other extensions like Github are able to be enabled/disabled per workspace when they have authentication providers in their extensions.

**Reply by sandy081** [reply]
Auth provider extensions cannot be disabled at workspace level. It is true for out of the box, GitHub and Microsoft Auth extensions.

**Reply by jpogran** [reply]
Forgive my confusion, but for example on my local machine both the GitHub Actions and Gitlens extension contribute authentication providers and both are allowed to be disabled per workspace.

I'm logged in as my GitHub user for both GitHub and GitHub Actions extensions:

<img width="489" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/272569/d86d2143-aed0-40d6-94d1-3eeff6cef4fc">

And can disable the extension for my workspace:

<img width="755" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/272569/d50a64e3-dbc1-477b-9473-e81ec703aff3">

I'm not currently logged into Gitlens, but can disable the extension for my workspace:

<img width="716" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/272569/8a87340a-7063-465a-88c5-30657a95619e">

What are those extensions doing differently that I am inside the Terraform extension?

<img width="663" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/272569/66d20781-22df-4269-aff4-6455b14fa185">


<img width="387" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/272569/d597bb5a-7994-4781-988a-c72c60bdef8b">

**Reply by tamuratak** [reply, marked as answer]
GitHub and GitHub Actions extensions doesn't have authentication providers. The built-in GitHub Authentication extension actually does, which has the following entry in `package.json`:

```json
  "contributes": {
    "authentication": ...
  }
```

https://github.com/microsoft/vscode/blob/0e17e5a38051148e8ed3655f1f603bac0c799553/extensions/github-authentication/package.json#L31

As you see, you cannot enable the built-in GitHub Authentication extension  at workspace.

**Comment by isidorn** [comment]
@sandy081 can you please help here?
